common/lib/xmodule/xmodule/js/src/capa/display.coffee

```diff
@@ -157,6 +159,75 @@ class @Problem
 
   # TODO add hooks for problem types here by inspecting response.html and doing
   # stuff if a div w a class is found
+          if response.progress_status != "done"
+            @maybeSetupTimer()
+
+  maybeSetupTimer: =>
+    if @el.find(".problem-timer").length
+      @$timer = @el.find('.problem-timer')
+      @$display = @el.find('.minutes-left')
+      @seconds_left = parseInt(@$timer.data('secondsLeft'), 10)
+      @seconds_before_warning = 60 * parseInt(@$timer.data('minutesBeforeWarning'), 10)
+      @submitted_before_time_expired = false
+
+      # The timer is initially hidden by CSS 
+      @$timer.show()
+
+      # Clear old timers, eg. if we press the submit button
+      if @timer_id
+        clearInterval(@timer_id)
+
+      # Sync every second
+      @timer_id = setInterval(@syncTimer, 1000)
+
+      # Initialize and show timer
+      @syncTimer()
+
+  syncTimer: =>
+    @seconds_left -= 1
+    if @seconds_left <= @seconds_before_warning
+      @showTimerWarning()
+
+    if @submitted_before_time_expired
+      @$timer
+        .removeClass("danger")
+        .text(gettext("Submitted"))
+
+    if @seconds_left <= 0
+      if not @submitted_before_time_expired
+        @$timer
+          .addClass("danger")
+          .text(gettext("Timer has expired"))
+      clearInterval(@timer_id)
+    else
+      @$display.text(@getDisplayText())
+
+  removeTimer: =>
+    @$(".problem-timer").remove()
+
+  showTimerWarning: =>
+    # if it's time to show a warning, show it
+    # both at the top of the problem and at the
+    # bottom. This is in case the user doesn't
+    # remember from a long problem that the problem
+    # is timed
+    if @$timer.length < 2
+      $clone = @el.find('.problem-timer').clone()
+      @el.find('.problem').after($clone.get(0))
+      # Re-find cached jQuery objects
+      @$timer = @el.find('.problem-timer')
+      @$display = @el.find('.minutes-left')
+      @$timer.addClass('danger')
+
+  getDisplayText: =>
+    s = @seconds_left
```

Comment:
It should really be "total_seconds_left". The variable represents the amount of seconds left for the user on the current problem (since the user could be jumping back into the middle of a timed problem). The backend knows the time the user truly started at, calculates the number of seconds remaining for the problem, then passes it to the front-end.